release/models/interfaces/openconfig-interfaces.yang

```diff
@@ -197,6 +204,18 @@ module openconfig-interfaces {
       name and subinterface index.";
   }
 
+  typedef interface-full-id {
+    type string {
+      pattern '([^.]+)(\.[0-9]+)?';
```

Comment:
For the most part I think we are probably ok on common "system-defined" physical/logical/virtual interface naming across network vendor implementations but this excludes any interface name that may carry a `.` unless the interface has no subinterface and encodes an integer post a single `.` (which is also legal linux netdev naming)

```
$ cat interfaces.txt 
et-0/0/0
et-0/0/0.0
et-0/0/0.1024
et-0/0/0:0.0
et-0/0/0:0.1024
ae0
ae0.0
ae0_0
TenGigabitEthernet0/0/0
TenGigabitEthernet0/0/0.0
mgmt0
mgmt0.0
Foo Bar Car
Foo Bar Car.1024
Foo Bar Car.1024.10
Foo/Bar/Car
Foo/Bar/Car:0
Foo/Bar/Car:0.100
Foo.Bar.Car/Jar
Foo.Bar.Car/Jar.100
Foo.Bar.Car/Jar.100
10/1/12
foo.1.2.3
foo.bar:0
foo.bar:0.0

$ cat interfaces.txt  | grep -vE "^([^.]+)(\.[0-9]+)?$" 
Foo Bar Car.1024.10
Foo.Bar.Car/Jar
Foo.Bar.Car/Jar.100
Foo.Bar.Car/Jar.100
foo.1.2.3
foo.bar:0
foo.bar:0.0
```

Maybe less of an issue for system-defined names but user-defined names (think virtual tunnels, etc..) now put a constraint on what the user can name an interface

Reply:
To entertain this idea a bit more, a robust way to be very permissive on interface names could be to replace the single string `interface-id` with a  strongly typed model representation with two leafs:
1. interface-name type string 
2. Optional, sub-interface-index type uint32

Thinking a bit about how to model this in yang, we could add/replace each leaf which uses the interface-id type with the above two leaves.  Today `typedef interface-id` is [present in 17 leafs](https://github.com/search?q=repo%3Aopenconfig%2Fpublic%20%22type%20oc-if%3Ainterface-id%22&type=code) which at a glance all relate some configuration to an interface.  

I am not sure it's worth the churn in the model to make a change like this and would like to hear what the issues are.

Reply:
Is there a problem with subinterface references that was not solved in #802 ?

Reply:
Am a bit confused - it seems what you are proposing is not much different than the `interface-ref-common` grouping:

https://github.com/openconfig/public/blob/master/release/models/interfaces/openconfig-interfaces.yang#L209-L235

But the problem trying to be solved here was the fact that we have a single leaf node key used in a list _somewhere_ that is either completely user-defined (made up objects that can differ all over each use) _or_ we use a typedef with some set of rules behind it to drive consistency in the list key naming per implementation.

Those "rules" are either gated by a pattern (which is likely too restrictive or too loose) or a very descriptive description stmt on an YANG unbounded string.

Now, this could be combined into the above grouping so that a common grouping of a list and child nodes (the ultimate distinct leafrefs) can be used and reapplied but this is primarily about ensuring that a key used to reference back to an interface (or subinterface combo) is consistent across all usage in each domain.

It's not about consistency across implementations but a set of rules within the same implementation at least.